Fix broken assert when joining empty strings.

saagarjha opened the pull request:
#### 9aab61cf8c873490106b14a93299b31bb8d188a8
<pre>
Fix broken assert when joining empty strings.

<a href="https://bugs.webkit.org/show_bug.cgi?id=291995">https://bugs.webkit.org/show_bug.cgi?id=291995</a>

Reviewed by Yusuke Suzuki.

This assert fails when joining an array of empty strings with an empty
string. Detect this case and bail out early.

* Source/JavaScriptCore/runtime/JSStringJoiner.cpp:
(JSC::joinStrings): Remove assert and early exit.

Canonical link: <a href="https://commits.webkit.org/294298@main">https://commits.webkit.org/294298@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/84a86a50ee93ac92430de0d409b0b71d78f4c1cf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101029 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20691 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10994 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106175 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51655 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21000 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29185 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76945 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33973 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104036 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16166 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91254 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57293 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15981 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9279 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51003 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/93696 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85887 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9337 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108531 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/99638 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28157 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20725 "Found 1 new test failure: http/tests/iframe-monitor/data-url-resource.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85913 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28519 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87454 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85451 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30172 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7898 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/22222 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16501 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28087 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33355 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/123263 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27899 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34342 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31219 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29457 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/JavaScriptCore/runtime/JSStringJoiner.cpp:
```diff
@@ -171,7 +171,8 @@ static inline String joinStrings(JSGlobalObject* globalObject, const WriteBarrie
 
-    ASSERT(joinedLength);
+    if (!joinedLength)
+        return StringImpl::empty();
```
Slightly nicer to say `emptyString()` here.

saagarjha replied:
Ah, neat. Let's use that then.

Constellation commented on Source/JavaScriptCore/runtime/JSStringJoiner.cpp:
```diff
@@ -171,7 +171,8 @@ static inline String joinStrings(JSGlobalObject* globalObject, const WriteBarrie
 
-    ASSERT(joinedLength);
+    if (!joinedLength)
+        emptyString();
```
return?

saagarjha replied:
Heh, yes. Thanks

saagarjha committed:
Fix broken assert when joining empty strings.

https://bugs.webkit.org/show_bug.cgi?id=291995

Reviewed by Yusuke Suzuki.

This assert fails when joining an array of empty strings with an empty
string. Detect this case and bail out early.

* Source/JavaScriptCore/runtime/JSStringJoiner.cpp:
(JSC::joinStrings): Remove assert and early exit.

Canonical link: https://commits.webkit.org/294298@main
